File: src-tauri/src/commands/xserver.rs
//! Tauri command surface for X server provisioning (#1052).
//!
//! Four commands back the frontend (settings toggle, connect-time provisioning,
//! and the Open Connections "X Servers" section — #1053):
//!
//! | Command | Purpose |
//! | --- | --- |
//! | [`x_server_status`] | Current server state (absent / adopted / running). |
//! | [`x_server_ensure`] | Ensure a usable server, or return a typed error. |
//! | [`x_server_stop`] | Stop the termiHub-managed server. |
//! | [`x_server_install_dependency`] | Install/guide the platform X dependency. |
//!
//! Progress is reported via the [`X_SERVER_PROGRESS_EVENT`] event, whose payload
//! ([`XServerProgress`]) mirrors the agent-deploy progress shape. The lifecycle
//! itself lives in [`crate::terminal::xserver::manager`] (#1049).

use std::sync::Arc;

use tauri::{AppHandle, State};

use crate::terminal::xserver::manager::XServerStatus as ManagedStatus;
use crate::terminal::xserver::{
    self, emit_progress, ConnectConsentRegistry, ConsentDecision, XServerError, XServerManager,
    XServerPlatform, XServerStatusReport,
};

/// Report the current local X server status without side effects.
#[tauri::command]
pub fn x_server_status(manager: State<'_, Arc<XServerManager>>) -> XServerStatusReport {
    xserver::current_status(&manager)
}

/// Ensure a usable local X server for the current platform.
///
/// Returns a coherent status on success (adopted external or managed running),
/// or a typed, actionable [`XServerError`] the UI can act on.
#[tauri::command]
pub async fn x_server_ensure(
    app: AppHandle,
    manager: State<'_, Arc<XServerManager>>,
) -> Result<XServerStatusReport, XServerError> {
    emit_progress(&app, "detect", "Checking for a local X server…", -1.0);

    let result = xserver::ensure_off_reactor(&app, manager.inner().clone()).await;

    match &result {
        Ok(outcome) => emit_progress(
            &app,
            "ready",
            outcome
                .report
                .message
                .as_deref()
                .unwrap_or("X server ready."),
            1.0,
        ),
        Err(err) => emit_progress(&app, "failed", &err.to_string(), 1.0),
    }
    result.map(|outcome| outcome.report)
}

/// Stop the termiHub-managed X server, if one is running.
///
/// Adopted external servers are never stopped — termiHub only shuts down what it
/// started.
#[tauri::command]
pub fn x_server_stop(
    app: AppHandle,
    manager: State<'_, Arc<XServerManager>>,
) -> Result<(), XServerError> {
    if !matches!(manager.status(), ManagedStatus::Running { .. }) {
        return Ok(());
    }
    manager.stop();
    emit_progress(&app, "stopped", "X server stopped.", 1.0);
    Ok(())
}

/// Install (or guide the install of) the platform's X dependency.
///
/// macOS runs a guided, consent-based XQuartz install (#1054): Homebrew when
/// present, otherwise actionable download guidance. Windows VcXsrv acquisition
/// (#1048) and Linux (never installs) still return their typed guidance so the
/// UI (#1053) has something concrete to show.
#[tauri::command]
pub async fn x_server_install_dependency(app: AppHandle) -> Result<(), XServerError> {
    emit_progress(&app, "install", "Preparing X dependency install…", -1.0);
    match XServerPlatform::current() {
        XServerPlatform::MacOs => finish(
            &app,
            xserver::macos::install_xquartz().await,
            "XQuartz is ready.",
        ),
        XServerPlatform::Windows => finish(
            &app,
            Err(XServerError::windows_provisioning_unavailable()),
            "",
        ),
        XServerPlatform::Linux => finish(&app, Err(XServerError::linux_install_unsupported()), ""),
    }
}

/// Emit the terminal progress event for an install `result` and return it, so the
/// success/failure tail is written once. Mirrors the `Ok/Err => emit` shape
/// [`x_server_ensure`] already uses.
fn finish(
    app: &AppHandle,
    result: Result<(), XServerError>,
    ok_message: &str,
) -> Result<(), XServerError> {
    match &result {
        Ok(()) => emit_progress(app, "ready", ok_message, 1.0),
        Err(err) => emit_progress(app, "failed", &err.to_string(), 1.0),
    }
    result
}

/// Deliver the user's reply to a connect-time X server download-consent prompt
/// (#1116), waking the SSH connect paused on the matching `id`.
///
/// Emitted by the `x-server-consent-needed` event; `decision` is `enable`
/// (consent — download/provision and remember it) or `notNow` (skip X
/// forwarding this connect). Returns `true` when a paused connect matched the
/// `id`, `false` when it was already resolved, cancelled, or unknown.
#[tauri::command]
pub fn x_server_connect_consent_reply(
    id: String,
    decision: ConsentDecision,
    registry: State<'_, Arc<ConnectConsentRegistry>>,
) -> bool {
    registry.resolve(&id, decision)
}

File: src/types/xserver.ts
/**
 * Types for the shared X server that termiHub manages (or adopts) for X11
 * forwarding. The backend tracks exactly one server at a time, so the UI shows
 * either zero or one server.
 */

/**
 * Lifecycle state of the shared X server.
 *
 * - `absent` — no X server is running or known.
 * - `adopted` — an external X server (not started by termiHub) is in use.
 * - `running` — a termiHub-managed X server is running.
 * - `failed` — the managed X server failed to start or crashed.
 */
export type XServerState = "absent" | "adopted" | "running" | "failed";

/** Platform the X server report was produced on. */
export type XServerPlatform = "windows" | "macOs" | "linux";

/**
 * Snapshot of the shared X server status as reported by the backend
 * (`x_server_status`). Optional fields are omitted when the backend has no
 * value (serialized from `Option::None`).
 */
export interface XServerStatusReport {
  /** Current lifecycle state of the server. */
  state: XServerState;
  /** Platform the report was produced on. */
  platform: XServerPlatform;
  /** X display number the server listens on (e.g. `0` for `:0`). */
  displayNumber?: number;
  /** Whether termiHub started and manages this server. */
  managed: boolean;
  /** Whether the platform dependency (e.g. VcXsrv) is installed. */
  dependencyAvailable?: boolean;
  /**
   * Number of live X11-forwarding sessions currently depending on this server.
   * Drives the Open Connections "· N sessions" detail; `0` when idle or absent.
   */
  sessionCount: number;
  /** Human-readable status detail, if any. */
  message?: string;
}

/**
 * Progress update emitted while the managed X server is being provisioned or
 * started (the `x-server-progress` event, fired during `x_server_ensure`).
 */
export interface XServerProgress {
  /** Machine-readable identifier for the current step. */
  step: string;
  /** Human-readable description of the current step. */
  message: string;
  /**
   * Completion fraction in the range `0`–`1`, or `-1` for an indeterminate
   * step (render a looping/indeterminate progress bar).
   */
  progress: number;
}

/**
 * A connect-time X server download-consent prompt (`x-server-consent-needed`
 * event, #1116). Emitted when opening an X11-forwarding SSH connection would
 * need to download the X dependency and the user has not consented yet. The
 * connect pauses until the frontend replies via `xServerConnectConsentReply`
 * with this `id`.
 */
export interface XServerConsentRequest {
  /** Opaque id correlating the prompt with the reply that resolves it. */
  id: string;
  /** Platform the connect is running on (tailors the consent copy). */
  platform: XServerPlatform;
}

/**
 * The user's reply to a connect-time consent prompt: `enable` downloads and
 * provisions (and is remembered), `notNow` skips X forwarding for this connect.
 */
export type XServerConsentDecision = "enable" | "notNow";

/**
 * Machine-readable classification of an X server provisioning failure. Drives
 * how the setup dialog recovers: `dependencyMissing` offers an install action;
 * the rest offer a plain retry with the human-readable `message`.
 */
export type XServerErrorKind =
  | "noLocalServer"
  | "dependencyMissing"
  | "serverUnreachable"
  | "launchFailed"
  | "unsupported";

/**
 * How the setup dialog should carry out the install action a `dependencyMissing`
 * error offers (#1309). The typed signal that replaced the earlier
 * `dependency === "Homebrew"` magic string; mirrors the Rust `InstallMode`.
 *
 * - `backend` — termiHub installs the dependency itself (`x_server_install_dependency`);
 *   any `installCommand` is shown for information only.
 * - `guidedTerminal` — the user runs `installCommand` in a terminal termiHub opens
 *   for them (the install has interactive prompts termiHub can't drive).
 * - `guidedExternal` — a prerequisite package manager is missing and isn't a
 *   terminal command either, so the user installs it from an external page/store
 *   termiHub opens, then retries (Windows winget-absent: App Installer, #1318).
 */
export type XServerInstallMode = "backend" | "guidedTerminal" | "guidedExternal";

/**
 * Typed error rejected by `x_server_ensure` / `x_server_install_dependency`.
 * Serialized from the tagged Rust `XServerError`; optional fields are omitted
 * when the backend has no value. `dependencyMissing` carries the missing
 * `dependency` plus optional install guidance.
 */
export interface XServerError {
  /** Failure classification. */
  kind: XServerErrorKind;
  /** Human-readable failure detail. */
  message: string;
  /**
   * Name of the missing dependency (only for `dependencyMissing`). Presentational
   * only — shown in the message and install-button label, never branched on.
   */
  dependency?: string;
  /**
   * How to run the install action (only for `dependencyMissing`, #1309): drives
   * whether the dialog executes {@link installCommand} in a terminal or triggers
   * the backend installer.
   */
  installMode?: XServerInstallMode;
  /** Human-readable hint on how to install the dependency. */
  installHint?: string;
  /** A concrete shell command that installs the dependency, if available. */
  installCommand?: string;
}

/**
 * Type guard for {@link XServerError}: an object carrying a string `kind` and a
 * string `message`. Use to narrow the `unknown` a rejected Tauri command
 * surfaces before reading typed guidance.
 */
export function isXServerError(e: unknown): e is XServerError {
  return (
    typeof e === "object" &&
    e !== null &&
    typeof (e as { kind?: unknown }).kind === "string" &&
    typeof (e as { message?: unknown }).message === "string"
  );
}
